content_handler/vulkan_surface_producer.cc

```diff
@@ -0,0 +1,370 @@
+      .imageType = VK_IMAGE_TYPE_2D,
+      .format = VK_FORMAT_B8G8R8A8_UNORM,
+      .extent = VkExtent3D{width, height, 1},
+      .mipLevels = 1,
+      .arrayLayers = 1,
+      .samples = VK_SAMPLE_COUNT_1_BIT,
+      .tiling = VK_IMAGE_TILING_OPTIMAL,
+      .usage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+      .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
+      .queueFamilyIndexCount = 0,
+      .pQueueFamilyIndices = nullptr,
+      .initialLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+  };
+
+  VkImage vk_image;
+  vk_result = VK_CALL_LOG_ERROR(vkCreateImage(
+      backend_context_->fDevice, &image_create_info, nullptr, &vk_image));
+  if (vk_result)
+    return nullptr;
+
+  VkMemoryRequirements memory_reqs;
+  vkGetImageMemoryRequirements(backend_context_->fDevice, vk_image,
+                               &memory_reqs);
+
+  uint32_t memory_type = 0;
+  for (; memory_type < 32; memory_type++) {
+    if ((memory_reqs.memoryTypeBits & (1 << memory_type)))
+      break;
+  }
+
+  VkMemoryAllocateInfo alloc_info = {
+      .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+      .pNext = nullptr,
+      .allocationSize = memory_reqs.size,
+      .memoryTypeIndex = memory_type,
+  };
+
+  VkDeviceMemory vk_memory;
+  vk_result = VK_CALL_LOG_ERROR(vkAllocateMemory(
+      backend_context_->fDevice, &alloc_info, NULL, &vk_memory));
+  if (vk_result)
+    return nullptr;
+
+  vk_result = VK_CALL_LOG_ERROR(
+      vkBindImageMemory(backend_context_->fDevice, vk_image, vk_memory, 0));
+  if (vk_result)
+    return nullptr;
+
+  const GrVkImageInfo image_info = {
+      .fImage = vk_image,
+      .fAlloc = {vk_memory, 0, memory_reqs.size, 0},
+      .fImageTiling = image_create_info.tiling,
+      .fImageLayout = image_create_info.initialLayout,
+      .fFormat = image_create_info.format,
+      .fLevelCount = image_create_info.mipLevels,
+  };
+
+  GrBackendRenderTargetDesc desc;
+  desc.fWidth = width;
+  desc.fHeight = height;
+  desc.fConfig = kSBGRA_8888_GrPixelConfig;
+  desc.fOrigin = kTopLeft_GrSurfaceOrigin;
+  desc.fSampleCnt = 0;
+  desc.fStencilBits = 0;
+
+  desc.fRenderTargetHandle = reinterpret_cast<GrBackendObject>(&image_info);
+
+  SkSurfaceProps props(SkSurfaceProps::InitType::kLegacyFontHost_InitType);
+
+  auto sk_surface = SkSurface::MakeFromBackendRenderTarget(context_.get(), desc,
+                                                           nullptr, &props);
+  if (!sk_surface) {
+    FTL_LOG(ERROR) << "MakeFromBackendRenderTarget Failed";
+    return nullptr;
+  }
+
+  uint32_t vmo_handle;
+  vk_result = VK_CALL_LOG_ERROR(vkExportDeviceMemoryMAGMA(
+      backend_context_->fDevice, vk_memory, &vmo_handle));
+  if (vk_result)
+    return nullptr;
+
+  mx::vmo vmo(vmo_handle);
+
+  size_t vmo_size;
+  vmo.get_size(&vmo_size);
+
+  FTL_DCHECK(vmo_size >= memory_reqs.size);
+
+  mx::eventpair retention_events[2];
+  auto mx_status =
+      mx::eventpair::create(0, &retention_events[0], &retention_events[1]);
```

Comment:
All this retention stuff should die, we don't use it anymore, instead you'll want to manage the acquire and release fences.  You probably don't need to observe these fences directly either.

Instead, it's sufficient to allocate two VkSemaphores for each image, call one the acquire fence and the other the release fence.  When you queue work for the image, have the queue wait for the release fence before issuing the work (unless it's the first time you are producing this image), and have the queue signal the acquire fence when the work completes.

Then export the semaphores' underlying event objects and give them to Mozart during the call to Present().